Add API Implementation

gmitch215 opened the pull request:
Merges into #1.

## Description

This PR does the following:
- Implements [CodeMC/API](https://github.com/CodeMC/API) for the Bot
  - Performs a RegEx grouping expression on the GitHub repository link, to be used to determine if a project is freestyle, and used as parameters for `username` and `jobName`
  - Creates a Nexus Repository (`/repository/username`), Jenkins User (`/job/username`), and Jenkins Job (`/job/username/job/jobName`) automatically 
  - Generates an arbitrary Nexus Password that is 32 characters long (see [generator.kt](https://github.com/CodeMC/API/blob/master/src/main/kotlin/io/codemc/api/generator.kt) for more details)
  - Performed Asynchronously with Kotlin Coroutines & Kotlin Serialization for better performance
- Updates the requirements in `config.json`
  - New `jenkins`, `nexus` and `database` sections are now required to be inputted. Similar to #4, specify the host URLs, with admin login details that the bot should use.
  - A new role list called `services` is available. This includes direct interaction with the current services via API. The sample includes the `Moderator` and `Administrator` roles.
  - See `config.json.sample` for more information.
- Adds a `codemc` Slash Command
  - This includes multiple user and administer functions related to the API. See the following linked issues for more information:
  - #7 
  - #8 
  - #9 
  - #10
- Removes the `project-url` parameter in the Accept command, calculated according to the `config.json` instead 
- Updates Gradle to the latest version, which is `8.9`.
- Adds more missing directories to `.gitignore`.

## API Implementation

The implementation details of the API are on the [repository](https://github.com/CodeMC/API), which performs [tests](https://github.com/CodeMC/API/actions/runs/10187861190/job/28182842850) against Jenkins and Nexus containers with similar configuration, and collects [code coverage](https://github.com/CodeMC/API/actions/runs/10187861190/job/28182975094). Thus, specific API changes must be made to the repository, which will then be updated here.

Andre601 commented on src/main/java/io/codemc/bot/CodeMCBot.java:
```diff
@@ -123,4 +139,8 @@ private void start() throws LoginException{
     public ConfigHandler getConfigHandler(){
         return configHandler;
     }
+
+    public Logger getLogger() {
+        return logger;
+    }
```
I'm not a fan of using a global getter for the Logger.
Instead should the classes using a Logger utilize the LoggerFactory to generate an instance with the right name applied. That way can we know in the logs where the message comes from, as with this getter, everything would be like `[CodeMCBot] Something something`, making it harder to know where this line came from.

gmitch215 replied:
Sounds good. I'll change that.

Andre601 commented on src/main/java/io/codemc/bot/JavaContinuation.java:
I assume this class is to utilize some Kotlin stuff in Java? I'm not experienced with Kotlin, so I don't quite understand the purpose of this class here to be honest.

gmitch215 replied:
Kotlin coroutines are like Java threads (but better) and are used for multithreading. They have special `suspend` functions to handle multithreading properly. This class was borrowed from [this tutorial](https://www.baeldung.com/kotlin/suspend-functions-from-java) since utilizing it here would need different kinds of callbacks.

Andre601 commented on src/main/java/io/codemc/bot/commands/CmdApplication.java:
```diff
@@ -38,11 +42,16 @@
 
 public class CmdApplication extends BotCommand{
+
+    public static final Pattern GITHUB_URL_PATTERN = Pattern.compile("^https://github\\.com/([^/]+)/([^/]+?)(?:\\.git)?(?:/.*)?$");
```
Shouldn't we use a more specific regex here? Like instead of `[^/]` which would allow any character that isn't `/` (including illegal ones like whitespaces), maybe `[a-zA-Z0-9-]` for the User part and `[a-zA-Z0-9-_.]` for the repository? Given those are values github accepts for user and repo names from what I could gather.

gmitch215 replied:
That sounds like a better idea. I'll change it.

Andre601 commented on src/main/java/io/codemc/bot/commands/CmdApplication.java:
```diff
@@ -119,8 +128,20 @@ public static void handle(CodeMCBot bot, InteractionHook hook, Guild guild, long
                 return;
             }
+
+            Matcher matcher = GITHUB_URL_PATTERN.matcher(repoLink);
+            if (!matcher.matches()) {
+                CommandUtil.EmbedReply.from(hook)
+                        .error("The provided Project URL did not match the pattern `https://ci.codemc.io/job/<user>/job/<project>`!")
+                        .send();
+                return;
+            }
```
This logic makes no sense. You check for a GitHub URL pattern but then report that the pattern didn't match the CI URL?
I assume this is a typo on your end here.

Andre601 replied:
Also, given projects may not necessarely be from GitHub Repositories (Only the username really has to be from GitHub), could this be somewhat limiting, if we do want to allow non-GitHub hosted projects, which we should for more freedom for the user.

gmitch215 replied:
> This logic makes no sense. You check for a GitHub URL pattern but then report that the pattern didn't match the CI URL? I assume this is a typo on your end here.

Yes, this is a typo. It's testing against the generated repository link, which in theory shouldn't fail.




> Also, given projects may not necessarely be from GitHub Repositories (Only the username really has to be from GitHub), could this be somewhat limiting, if we do want to allow non-GitHub hosted projects, which we should for more freedom for the user.

I am in favor of a GitHub-only approach since other Git providers like GitLab (along with self-hosted) are marketed for companies or individuals with the resources to not need the service. That being said, I can change the API/Logic here to add support for GitLab. Never had any experience outside of those two though.

Andre601 replied:
My main issue with a GitHub-only aproach here is the fact that GitHub can be (And is) predatory in certain aspects. Just see the Copilot fiasco, where it used code from Repositories without providing credit, violating dozens of licenses. All without any opt-out option for the dev themself to not suffer this problem.

A user may choose a different host, or even host an instance themself for various reasons...
Tho I guess we can look at this in the future as most if not all current requests seem to be GitHub hosted projects.

gmitch215 replied:
If they all are GitHub projects, I think we should just drop support for the rest. 

GitLab has its own version of copilot and is probably more predatory since it's marketed for more industrial-based organizations who can pay for all that stuff.

And, anyone with enough money to pay for their own Git server has the funds for their own nexus and CI instances, since most who do are worried about control and privacy the most. 

I would only add support if we have existing users in GitHub. Would recommend a simple discord announcement detailing the status update and such.

Andre601 replied:
So, what is the stand with this one now?

gmitch215 replied:
Universal Git Support is possible and implemented. See [network.kt](https://github.com/CodeMC/API/blob/master/src/main/kotlin/io/codemc/api/network.kt#L49) for more details.

Andre601 commented on src/main/java/io/codemc/bot/commands/CmdApplication.java:
```diff
@@ -169,11 +190,45 @@ public static void handle(CodeMCBot bot, InteractionHook hook, Guild guild, long
+            String password = Generator.createPassword(24);
+
+            CompletableFuture<Boolean> nexus = new CompletableFuture<>();
+            nexus.handle((v, ex) -> {
```
Personal preference, but I feel like the value should have a meaningful name instead of just `v`, such like `result`. It's not a big deal tho.

Andre601 replied:
Also, any particular reason for using handle here and not handleAsync or supply(Async)?
Is it related to the JavaContinuation you added?

gmitch215 replied:
> Personal preference, but I feel like the value should have a meaningful name instead of just `v`, such like `result`. It's not a big deal tho.

Agreed.

> Also, any particular reason for using handle here and not handleAsync or supply(Async)? Is it related to the JavaContinuation you added?

I rarely have used the `CompletableFuture` class since most of my multithreading knowledge is linked to Kotlin, which handles it more cleanly (IMO). If it's better to use that method, then I'll switch it.

Andre601 replied:
I assume it's better, given supply/handleAsync do stuff asynchronously using a separate thread... Tho, this leaves the possible issue that the execution continues here before the futures are done with their work, so unless this is handled later, which from what I understand it is with the JavaContinuation thingy, could this be a headache to handle... But handle itself seems to be sync from what I gather here...
Tho, I myself lack the major knowledge of CF to give good advice here.

gmitch215 replied:
It doesn't matter if Jenkins executes parallel to Nexus (and would actually be preferable for performance), so it sounds like switching is favorable.

Andre601 commented on src/main/java/io/codemc/bot/commands/CmdApplication.java:
```diff
@@ -169,11 +190,45 @@ public static void handle(CodeMCBot bot, InteractionHook hook, Guild guild, long
+                        .error("Failed to create Nexus Repository!")
+                        .send();
+
+                    bot.getLogger().error("Failed to create Nexus Repository for {}!", username, ex);
```
As mentioned earlier should this not use a global logger, but instead a local logger instance created with the LoggerFactory (`Logger logger = LoggerFactory.getLogger(CmdApplication.class)`)

gmitch215 replied:
Will fix.

Andre601 commented on src/main/java/io/codemc/bot/JavaContinuation.java:
```diff
@@ -0,0 +1,40 @@
+        return Dispatchers.getIO();
+    }
+
+    public static <T> JavaContinuation<T> create(CompletableFuture<T> future) {
+        return new JavaContinuation<>(future);
+    }
```
Iirc should public static methods be located above public ones based on common Java conventions. I can be wrong here...

gmitch215 replied:
From what I quickly searched, some agree and others say that it should follow the rest of the codebase. So I'll just move it.

Andre601 commented on src/main/java/io/codemc/bot/commands/CmdApplication.java:
```diff
@@ -220,21 +272,13 @@ public void withHookReply(InteractionHook hook, SlashCommandEvent event, Guild g
                     return -1L;
                 }
             });
```
Looks like I forgot to update these lines in my own PR. Could you correct this to make it the same as used in the Deny class?
```java
long messageId = event.getOption("id", -1L, OptionMapping::getAsLong);
```

gmitch215 replied:
Will do.

Andre601 commented on src/main/java/io/codemc/bot/listeners/ModalListener.java:
```diff
@@ -247,15 +247,7 @@ public void onModalInteraction(@NotNull ModalInteractionEvent event){
-                }
-                
-                CmdApplication.handle(bot, hook, guild, messageId, text, accepted);
+                CmdApplication.handle(bot, hook, guild, messageId, null, accepted);
```
I believe this part of the modal listener isn't even used anymore, as it was only used in context menus, so this change is not required (Although you are free to remove the entry `case` section if you want).

gmitch215 replied:
Sounds good. I'll remove it.

Andre601 reviewed (changes requested):
A lot of things to discuss, with some changes I would apreciate to be made.

gmitch215 commented:
Updates the implementation from [CodeMC/API](https://github.com/CodeMC/API/compare/0b2d3d3...master). Unit tests pass.

I will make a separate PR implementing a `CmdRemove` that goes into #1 separately.

gmitch215 commented:
GitHub is weird.

Andre601 commented on src/main/java/io/codemc/bot/commands/CmdRemove.java:
```diff
@@ -0,0 +1,122 @@
+        this.allowedRoles = bot.getConfigHandler().getLongList("allowed_roles", "commands", "service");
+
+        this.options = List.of(
+                new OptionData(OptionType.USER, "user", "The user that is having their status revoked.").setRequired(true),
+                new OptionData(OptionType.STRING, "username", "The Jenkins/Nexus username of the user. If blank, defaults to their server nickname.").setRequired(false)
```
I am not a fan of this aproach here.
While I can see the reasons am I not too happy to use the Discord user as the main source of finding and handling the CI user.
It would be better to just make the username option required too.
```suggestion
                new OptionData(OptionType.USER, "user", "The user that is having their status revoked.").setRequired(true),
                new OptionData(OptionType.STRING, "username", "The Jenkins/Nexus username of the user. If blank, defaults to their server nickname.").setRequired(true)

```

Andre601 commented on src/main/java/io/codemc/bot/commands/CmdRemove.java:
```diff
@@ -0,0 +1,122 @@
+            return;
+        }
+
+        String username = event.getOption("username", user.getNickname(), OptionMapping::getAsString);
```
As stated earlier, not a fan of having the Discord user be used as main source of getting the CI User. After all could their Discord name not match their GitHub name.

So if the above is changed should this here be altered too:
```suggestion
        String username = event.getOption("username", null, OptionMapping::getAsString);
        
        if (username == null || username.isEmpty()) {
            CommandUtil.EmbedReply.from(hook).error("Invalid CI User provided!").send();
            return;
        }
```

Andre601 commented on src/main/java/io/codemc/bot/commands/CmdRemove.java:
```diff
@@ -0,0 +1,122 @@
+        NexusAPI.deleteNexus(username, JavaContinuation.create(deleted));
+
+        guild.removeRoleFromMember(user, authorRole)
+                .reason("[Access Request] Author Status revoked by " + member.getUser().getAsTag())
+                .queue(
+                        v -> CommandUtil.EmbedReply.from(hook)
+                                .success("Revoked Author Status from " + member.getUser().getEffectiveName() + "!")
```
Wrong member used here.
```suggestion
                .reason("[Access Request] Author Status revoked by " + user.getUser().getAsTag())
                .queue(
                        v -> CommandUtil.EmbedReply.from(hook)
                                .success("Revoked Author Status from " + user.getUser().getEffectiveName() + "!")

```

Andre601 commented on src/main/java/io/codemc/bot/commands/CmdRemove.java:
```diff
@@ -0,0 +1,122 @@
+        this.name = "remove";
+        this.help = "Revoke a user's Author status.";
+
+        this.allowedRoles = bot.getConfigHandler().getLongList("allowed_roles", "commands", "service");
```
Please name the option properly to match the actual command name.
```suggestion
        this.allowedRoles = bot.getConfigHandler().getLongList("allowed_roles", "commands", "remove");

```

gmitch215 replied:
I named it `service` because I wanted it to be used as the permission for #8. If you think that should be `accept` instead, then I'll change it.

Andre601 replied:
The overall structure mimics the commands themself, so that it is easier for a new user to understand it, if they ever want to use the bot themself.

Perhaps, to prepare for #8 could we make the remove command be a sub-command of `/service` or something similar, just like application is a main command with sub-commands.
You could then have the `service` config option while having remove and alike...

Tho, maybe there's a better aproach than /service for the command then? Like /jenkins? Would fit the overall purpose of the command (Jenkins-related stuff)... Tho maybe not the nexus... Would apreciate your input on this.

gmitch215 replied:
I wouldn't name it `/jenkins` since it would manage the Nexus state.

I'm assuming subcommands can have differing permissions, so adding it to a `/service` root command could be nice.

Perhaps a `/jenkins` or `/nexus` command could suffice for #9? My main goal would be for users to anonymously read jenkins/nexus state from Discord. I'm unsure how authentication would work for write access, if desired.

Andre601 commented on src/main/resources/config.json:
```diff
@@ -14,6 +14,7 @@
     },
     "commands": {
       "application": [],
+      "service": [],
```
```suggestion
      "remove": [],
```

Andre601 replied:
```suggestion
      "codemc": [],
```

Andre601 commented on src/main/resources/config.json.sample:
```diff
@@ -26,6 +26,10 @@
         659568973079379971,
         1233971297185431582
       ],
+      "service": [
```
```suggestion
      "remove": [
```

Andre601 replied:
```suggestion
      "codemc": [
```

Andre601 reviewed (changes requested):
A few things that should be changed.
Main concern is the reliance on the Discord user, with the jenkins username only being optional. Not every user matches their GitHub and Discord names, and using something like the user's nickname is not a good aproach here.

gmitch215 commented:
Added recommended changes to the `/codemc` command. Includes #7, #8, and #9. 

The implementation for #8 includes using Kotlin Coroutines for validating all the registered Jenkins users, aiming to help speed up validation and API creation. A quick look at the public API endpoint shows ~101 authors, so this will be useful.

gmitch215 commented:
Added a `/codemc changepassword` subcommand, along with `/codemc unlink` and a few other utilities.

Andre601 commented on src/main/java/io/codemc/bot/commands/CmdCodeMC.java:
```diff
@@ -0,0 +1,486 @@
+            }
+
+            DatabaseAPI.removeUser(username);
+            CommandUtil.EmbedReply.from(hook).success("Unlinked " + target.getUser().getAsTag() + " from their Jenkins/Nexus account!").send();
```
```suggestion
            CommandUtil.EmbedReply.from(hook).success("Unlinked Discord User " + target.getUser().getEffectiveName() + " from their Jenkins/Nexus account!").send();

```

Andre601 commented on src/main/java/io/codemc/bot/utils/APIUtil.java:
```diff
@@ -0,0 +1,115 @@
+    public static void createNexus(InteractionHook hook, String username, String password) {
+        CompletableFuture<Boolean> nexus = new CompletableFuture<>();
+        nexus.whenCompleteAsync((success, ex) -> {
+            if(ex != null || (success != null && !success)){
+                if (hook != null)
+                    CommandUtil.EmbedReply.from(hook)
+                            .error("Failed to create Nexus Repository!")
+                            .send();
+
+                LOGGER.error("Failed to create Nexus Repository for {}!", username, ex);
```
The logger message could throw afaik, because `ex` can be null by the current `if` check (if ex is not null OR success is not null and false).

I would separate the success check into its own if check and return a different message here for the exception (i.e. that an error was raised while creating the repository).

gmitch215 replied:
Sounds good.

Andre601 commented on src/main/java/io/codemc/bot/utils/APIUtil.java:
```diff
@@ -0,0 +1,115 @@
+    public static void changePassword(InteractionHook hook, String username, String newPassword) {
+        CompletableFuture<Boolean> changeNexusPassword = new CompletableFuture<>();
+        changeNexusPassword.whenCompleteAsync((nexusSuccess, nexusEx) -> {
+            if (nexusEx != null || (nexusSuccess != null && !nexusSuccess)){
+                if (hook != null)
+                    CommandUtil.EmbedReply.from(hook)
+                            .error("Failed to change Nexus Repository password!")
+                            .send();
+
+                LOGGER.error("Failed to change Nexus Repository password for {}!", username, nexusEx);
```
Same stuff I mentioned earlier.

Andre601 commented on src/main/java/io/codemc/bot/utils/APIUtil.java:
```diff
@@ -0,0 +1,115 @@
+
+            CompletableFuture<Boolean> changeJenkinsPassword = new CompletableFuture<>();
+            changeJenkinsPassword.whenCompleteAsync((jenkinsSuccess, jenkinsEx) -> {
+                if (jenkinsEx != null || (jenkinsSuccess != null && !jenkinsSuccess)){
+                    if (hook != null)
+                        CommandUtil.EmbedReply.from(hook)
+                                .error("Failed to change Jenkins password!")
+                                .send();
+
+                    LOGGER.error("Failed to change Jenkins password for {}!", username, jenkinsEx);
```
Same issue I mentioned before.

gmitch215 commented on src/main/java/io/codemc/bot/commands/CmdCodeMC.java:
```diff
@@ -0,0 +1,486 @@
+            }
+
+            String jenkinsUrl = bot.getConfigHandler().getString("jenkins", "url");
+            String username = job.split("/")[0];
+            String jobName = job.split("/")[1];
```
I'll add validation to be safe. It only needs to contain a `/`.

gmitch215 commented on src/main/java/io/codemc/bot/commands/CmdCodeMC.java:
```diff
@@ -0,0 +1,486 @@
+                String type = ((JsonPrimitive) info.get("type")).getContent();
+
+                MessageEmbed embed = CommandUtil.getEmbed()
+                        .setTitle(user, nexusUrl + "/#browse/browse:" + repository)
+                        .setAuthor(user, nexusUrl + "/#browse/browse:" + repository)
```
Gotcha.

gmitch215 commented on src/main/java/io/codemc/bot/commands/CmdCodeMC.java:
```diff
@@ -0,0 +1,486 @@
+                }
+            });
+            NexusAPI.deleteNexus(username, JavaContinuation.create(deleted));
+            DatabaseAPI.removeUser(username);
```
I had merging issues when I moved development between computers and must've missed that I added the statement twice.

gmitch215 commented on src/main/java/io/codemc/bot/commands/CmdCodeMC.java:
```diff
@@ -0,0 +1,486 @@
+            CommandUtil.EmbedReply.from(hook).success("Successfully validated " + count.get() + " User(s)").send();
+        }
+
+        private static void validate(CoroutineScope scope, String username, Runnable callback) {
```
It is `private` because it is only used in the file and `static` because an instance isn't needed. Usually I have both in there on these conditions regardless of where the method is nested. Would you recommend removing `static`?

gmitch215 commented on src/main/java/io/codemc/bot/commands/CmdCodeMC.java:
```diff
@@ -0,0 +1,486 @@
+                    if (info == null || info.isEmpty())
+                        APIUtil.createNexus(null, username, password);
+
+                    callback.run();
```
I originally thought of having a separate implementation for singular validation, then just forgot to remove the redundancy.

Andre601 commented on src/main/java/io/codemc/bot/utils/APIUtil.java:
I just noticed you use `{}` in `EmbedReply` messages.
These won't work with String.format iirc, so they need to be changed to proper placeholder values based on their type.

Only doing this comment here.

gmitch215 commented:
That should be all of them. I also upgraded Gradle to the most recent version.

Andre601 commented on src/main/java/io/codemc/bot/commands/CmdCodeMC.java:
```diff
@@ -0,0 +1,487 @@
+
+            this.options = List.of(
+                    new OptionData(OptionType.STRING, "username", "The Jenkins/Nexus username of the user.").setRequired(true),
+                    new OptionData(OptionType.USER, "discord", "The discord user that is having their status revoked. Leave blank if the user is no longer in the server.")
```
Description is too long (Can't be more than 100 characters).
My change:
```suggestion
                    new OptionData(OptionType.USER, "discord", "ID of Discord user having their status revoked. Leave blank for users no longer in the server.")

```

Andre601 commented on src/main/java/io/codemc/bot/commands/CmdCodeMC.java:
```diff
@@ -0,0 +1,487 @@
+            this.allowedRoles = bot.getConfigHandler().getLongList("allowed_roles", "commands", "codemc");
+
+            this.options = List.of(
+                    new OptionData(OptionType.STRING, "username", "The target username to validate. If left blank, validates all existing Jenkins users.")
```
```suggestion
                    new OptionData(OptionType.STRING, "username", "Target Username to validate. When left blank, validates all existing Jenkins Users.")
```

gmitch215 commented on build.gradle:
```diff
@@ -42,6 +42,7 @@ dependencies {
     implementation group: 'pw.chew', name: 'jda-chewtils-commons', version: '2.0'
     implementation group: 'pw.chew', name: 'jda-chewtils-command', version: '2.0'
     implementation group: 'org.spongepowered', name: 'configurate-gson', version: '4.1.2'
+    implementation group: 'io.codemc.api', name: 'codemc-api', version: '1.0.0'
```
Most likely the fact that Kotlin is cached locally. The first one is JetBrains's SQL API for Kotlin, the second being the multithreading library, and the third being the JSON parser. You should be able to view the proper versions in the repository.

gmitch215 commented:
Tested everything on a Bot with a Docker Setup.

A few notes:
- There were a few bugs in the [API](https://github.com/CodeMC/API/compare/bef6a847c4b8344c56f3a7bfc56e4d0802135629...master) regarding bot-specific actions, like the `validate` command where it could create inconsistencies between Jenkins and Nexus passwords.
- Added more logging for future debugging and just general information.
- `submit` now denies users that apply with the same name.
  - It also appears JDA makes fields required by default, meaning optionality must be explicit.
- There were a few bugs regarding typos in configuration paths, along with a small bug for `lacksRole` in `ButtonListener`.
- The `Deny` button has a literal `"Reason not Provided"` as a reason, since making it null replaces it with the Jenkins URL for success
- The `remove` command uses the database to remove the author role from the discord user, removing it as a parameter.
- I [removed](https://github.com/CodeMC/API/compare/bef6a847c4b8344c56f3a7bfc56e4d0802135629...master) a lot of `suspend` modifiers in the API, which allows for better interoperability with Java. I also noticed a lack of faster runtime for having coroutines in the `validate` command where it validates all Jenkins users, so I also removed it there.
- I fixed the dependencies issue by adding all of the necessary dependencies, along with the CodeMC API itself.
